cip-multiout-reverse.md

```diff
@@ -0,0 +1,138 @@
+
+5. Implement all of this in a way that does not require the network to fork.
+
+### Proposed Solution:
+
+1. Implement a background task that will scan existing blocks and, when new blocks come in, scan them, and analyze all Multi-Out (Same) transactions. When this task finds a Multi-Out (Same) transaction it should add each recipient account ID and the transaction ID of the current transaction to a new database table, which is defined as follows:
+
+| Column Name | Data Type | Description | Example Value |
+|-|-|-|-|
+| recipient_id | BIGINT (64-bit unsigned long) | The account ID of the recipient of a Multi-Out (Same) transaction | 7009665667967103287 |
+| transaction_id | BIGINT (64-bit unsigned long) | The transaction ID of the Multi-Out (Same) transaction in which this account was paid | 3631659512270044993 |
+
+For example, if the service process a Multi-Out with transaction ID `TransactionA` and accounts `AccountA`, `AccountB` and `AccountC` are paid an amount, and then processes a Multi-Out Same with transaction ID `TransactionB` where accounts `AccountA` and `AccountD` are paid, it would add the rows:
+
+| recipient_id | transaction_id |
+|-|-|
+| AccountA | TransactionA |
+| AccountB | TransactionA |
+| AccountC | TransactionA |
+| AccountA | TransactionB |
+| AccountD | TransactionB |
+
+2. When a user requests their transaction list, search the table for all rows where `recipient` is equal to the user's account ID, which would in this example for `AccountA` return:
+
+| recipient_id | transaction_id |
+|-|-|
+| AccountA | TransactionA |
+| AccountA | TransactionB |
+
+3. Lookup the transactions found and add them to the list of regular transactions obtained from the transactions table in the database.
+
+4. When displaying the list of transactions, display the incoming amount as the amount that `AccountA` was paid in the Multi-Out (Same) transaction. There will only be one entry in the multi-out recipients as accounts cannot be sent to twice in the same Multi-Out (Same) transaction.
+
+### Impact on database storage
+
+Each row in the new table consumes 16 bytes, and there is one per recipient per transaction.
+
+Therefore the maximum amount of additional space consumed per Multi-Out transaction is `16*64 = 1024` bytes, which is `1024/1201 = 85%` more than was previously consumed per Multi-Out transaction.
+
+Additionally, the maximum amount of additional space consumed per Multi-Out Same transaction is `16*128 = 2048` bytes, which is `2048/1201 = 171%` more than was previously consumed per Multi-Out transaction.
+
+Additionally, the new space consumption decrease (compared to not having Multi-Out transactions) is now `(11264-(1201+1024))/11264 = 80.2%` (down from `89.3%` without the new table).
+
+Additionally, the new space consumption decrease (compared to not having Multi-Out Same transactions) is now `(22528-(1201+2048))/22528 = 85.6%` (down from `94.7%` without the new table).
```

Comment:
As far as I understood, we'll have an increase in db space usage, although compared to pre-CIP-4 implementation, we're still quite efficient in terms of space, right?